File: week_8_and_9/3_Ruby/fizz_buzz/my_solution.rb
# U3.W8-9: 


# I worked on this challenge [by myself].

# 2. Pseudocode

# define a method named super_fizzbuzz
# pass an argument to the method above
# If the array value is a multiple of 3, return "Fizz"
# If the array value is a multiple of 5, return "Buzz"
# If the array value is a multiple of 3 and 5, return "FizzBuzz"
# else return the original array

# 3. Initial Solution

def super_fizzbuzz(array)
  # Your code goes here!
  new_array = []
  array.each do |i|
    if i % 3 == 0
      if i % 5 == 0
        new_array << "FizzBuzz"
      else
        new_array << "Fizz"
      end
    elsif i % 5 == 0
      new_array << "Buzz"
    else
      new_array << i
    end
  end
  new_array
end


# 4. Refactored Solution

def super_fizzbuzz(array)
  array.map do |i|
    if i % 3 == 0
      i % 5 == 0 ? "FizzBuzz": "Fizz"
    elsif i % 5 == 0
      "Buzz"
    else
      i
    end
  end
end

# 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE

p super_fizzbuzz([1,2,3])  # => [1, 2, "Fizz"]
p super_fizzbuzz([1,2,5])  # => [1, 2, "Buzz"]
p super_fizzbuzz([1,2,15]) # => [1, 2, "FizzBuzz"]
p super_fizzbuzz([30, 9, 20, 1]) # => ["FizzBuzz", "Fizz", "Buzz", 1]

# 5. Reflection 

# I had previously worked on this exercise at socrates.devbootcamp.com but I was able to refactor
# the solution, further.

File: week_8_and_9/2_JavaScript/grocery_list/my_solution.js
// U3.W8-9: 


// I worked on this challenge [by myself].

// 1. User Stories

// As a user, I want to be able to create a new grocery list.
// As a user, I want to be able to add items to the list.
// As a user, I want to be able to update the quantity for a given item.
// As a user, I want to be able to remove items from the list.
// As a user, I want to be able to show the items in the list.

// 2. Pseudocode

/*
- Create an object "GroceryList"
- Add items to the list in form of a Hash
- Create a function "show" to show items in a list.
- Create a function "add" to add an item to the list.
- Create a function "update" to update the quantity for a given item.
- Create a function "remove" to remove an item from the list.

*/

// 3. Initial Solution

var GroceryList = {
  items: {
    eggs: 12,
    milk: 2,
    pears: 8,
    bananas: 6 
  },

// ---------------- View all the items in the list -----------------
  show: function(){
    console.log("Grocery List:\n");
    for (var key in this.items) {
        console.log(this.items[key] + " " + key);
    }
    console.log();
  },

// ---------------- Add an item to the list ------------------------
  add: function(item, quantity){
    this.items[item] = quantity;
    console.log("You added "+ quantity + " " + item +". Your Grocery list has been updated.\n");
  },

// ---------------- Update the quantity of an item in the list -----
  update: function(item, quantity){
    this.items[item] = quantity;
    console.log("You updated the quantity of " + item +". Your Grocery list has been updated.\n");
  },

// ---------------- Remove an item from the list -------------------
  remove: function(item){
    delete this.items[item];
    console.log("You removed "+ item +". Your Grocery list has been updated.\n");
  }
};


// 4. Refactored Solution






// 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE

GroceryList.show();

GroceryList.add("apples", 10);
GroceryList.show();

GroceryList.update("pears", 5);
GroceryList.show();

GroceryList.remove("milk");
GroceryList.show();

// 5. Reflection 

/*
The concepts of Javascript faded as I was more inclined towards Ruby for past few weeks. I used Codecademy
to refresh my understanding of Javascript and used the link below to finish this exercise:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Introduction_to_Object-Oriented_JavaScript
*/
